### PR TITLE
refactor(pubsub): split SubscriptionSession [3]

### PR DESCRIPTION
**Pull request by coryan**
Copy the functionality to perform (limited) flow control to a separate
class. For now, this class is only used in tests. A future PR will
replace `pubsub_internal::SubscriptionSession` with a class that uses
this new component.

Part of the work for #4970 and for #4645 

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4994)

### CODE REVIEW

File: google/cloud/pubsub/internal/subscription_flow_control_test.cc [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+  }
+
+  void SaveAcks(google::pubsub::v1::ReceivedMessage const& m) {
+    std::unique_lock<std::mutex> lk(acks_mu_);
+    acks_.push_back(m.ack_id());
+    lk.unlock();
+    acks_cv_.notify_one();
+  }
+
+  void WaitAcks() {
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_flow_control_test.cc:74]: Opt: Perhaps just use `WaitAcksCount(1)`?

File: google/cloud/pubsub/internal/subscription_flow_control_test.cc [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+      .WillOnce([&](std::string const& ack_id, std::size_t) {
+        EXPECT_THAT(ack_id, "ack-0-0");
+        return make_ready_future(Status{});
+      });
+  EXPECT_CALL(*mock, NackMessage)
+      .WillOnce([&](std::string const& ack_id, std::size_t) {
+        EXPECT_THAT(ack_id, "ack-1-0");
+        return make_ready_future(Status{});
+      });
+  // There may be a few of these calls during shutdown, ignore them, we have
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_flow_control_test.cc:126]: Nit: s/, ignore them, we/, so ignore them. We/ perhaps?

File: google/cloud/pubsub/internal/subscription_flow_control.cc [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  if (!response) {
+    shutdown_manager_->FinishedOperation("OnPull");
+    shutdown_manager_->MarkAsShutdown(__func__, std::move(response).status());
+    Shutdown();
+    return;
+  }
+  if (shutdown_manager_->FinishedOperation("OnPull")) {
+    std::vector<std::string> ack_ids(
+        response->mutable_received_messages()->size());
+    std::size_t total_size = 0;
```
Comment (devbww) [inline google/cloud/pubsub/internal/subscription_flow_control.cc:102]: Unused?

Comment (coryan) [reply google/cloud/pubsub/internal/subscription_flow_control.cc:102]: Fixed.

File: google/cloud/pubsub/internal/subscription_flow_control_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_flow_control_test.cc]: Fixed.

File: google/cloud/pubsub/internal/subscription_flow_control_test.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_flow_control_test.cc]: Done.

File: google/cloud/pubsub/internal/subscription_flow_control_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/pubsub/internal/subscription_flow_control_test.cc]: Or at least "changed", let me know if you think it needs further improvement.

**Pull request merged**